### ISSUE TITLE
ColorVariable should flag uses of rgb/rgba/hsl/hsla

### ISSUE THREAD

**Issue by cH40z-Lord** [brigade/scss-lint#488]
Today I've discovered a strange inconsistency between color definitions. SCSS-Lint is throwing a warning whenever you try to use a color definition like `#fff` outside of a SCSS variable. However, this is not true for colors defined via `rgb()` and `rgba()`, which feels kinda strange and not what I expect.

```css
.foo {
  // warning: foo.scss:2 [W] ColorVariable: Color literals like `#fff` should only be used in variable declarations; they should be referred to via variable everywhere else.
  backgorund-color: #fff;
}

.bar {
  // no warning/error
  background-color: rgba(255, 255, 255, .5);
  //background-color: rgb(255, 255, 255);
}
```

I'd love to have an option for this as well so we can improve our SCSS code even more. What do you think?

**Comment by apfelbox** [comment]
Maybe regression by #402?

Here is my understanding / proposal of the `ColorVariable` linter:

| Type                                            | Code Examples                    | SHOULD report? | DOES CURRENTLY report? |
| ----------------------------------------------- | -------------------------------- | -------------- | ---------------------- |
| **Regular color definition**                    | `color: red;`                    | yes            | yes ✓                  |
|                                                 | `color: #fff;`                   | yes            | yes ✓                  |
|                                                 | `color: #ffffff;`                | yes            | yes ✓                  |
|                                                 | `color: rgb(0, 100, 200);`       | yes            | **no ✗**               |
|                                                 | `color: rgb(0%, 50%, 80%);`      | yes            | **no ✗**               |
|                                                 | `color: rgba(0, 0, 0, .5);`      | yes            | **no ✗**               |
|                                                 | `color: hsl(0, 100%, 50%);`      | yes            | **no ✗**               |
|                                                 | `color: hsla(0, 100%, 50%, .5);` | yes            | **no ✗**               |
| **Special CSS keywords**                        | `color: currentColor;`           | no             | no ✓                   |
|                                                 | `color: inherit;`                | no             | no ✓                   |
| **Resetting a color**                           | `color: transparent;`            | no             | no ✓                   |
| **Directly using a variable**                   | `color: $red;`                   | no             | no ✓                   |
| **Transforming a variable value in a function** | `color: rgba($red, .5);`         | no             | no ✓                   |
|                                                 | `color: lighten($red, 5%);`      | no             | no ✓                   |


Tested with scss-lint 0.39.0 and sass 3.4.14

**Comment by sds** [comment]
I wouldn't say this is a regression due to #402, I would say we just never considered treating `rgb`/`rgba`/`hsl`/`hsla` as literal colors, but I suppose they would qualify.

I'm open to adding support to `ColorVariable` to disable color functions as well.

**Comment by apfelbox** [comment]
All right, thanks. I was under the impression that it was supported before, but then I don't use color definitions other than hex values regularly.

Support would be very nice. Does the support matrix above look good to you?

**Comment by sds** [comment]
@apfelbox your matrix looks reasonable to me.

**Comment by sds** [comment]
Implemented in #569.

**Comment by darbyw** [comment]
I think this change might have made it so that the following is flagged as a `ColorVariable` violation:

```
$panel-shadow-color: rgba(0, 0, 0, 0.3);
```

Is this intended?  This is not covered by the matrix above, but it seems like it should pass since we're assigning the color to a variable to be used elsewhere (which is what `ColorVariable` is intended to promote).

**Comment by apfelbox** [comment]
Variable definitions should never be flagged. Only the direct usage of colors, *without* storing it in a variable first.

**Comment by darbyw** [comment]
@apfelbox I see - there might be a regression then.  We had to disable the rule to get our build to pass:

```
// scss-lint:disable ColorVariable
$dropdown-menu-shadow:          rgba(0, 0, 0, 0.2);
$panel-shadow:                  rgba(0, 0, 0, 0.3);
// scss-lint:enable ColorVariable
```

Let me know if there's any further information I can provide!

**Comment by peacefulrespite** [comment]
@darbyw after exploring this more, the lint can be addressed with putting the `RGB` part of the color into a variable and then using that in your variable definition w/ the alpha. For example:

``` scss
$white: #FFFFFF;
$dropdown-menu-shadow: rgba($white, 0.2);
$panel-shadow: rgba($white, 0.3);
```

will satisfy the lint and should work as expected.

**Comment by darbyw** [comment]
Thanks @peacefulrespite, that makes sense...

Although per @apfelbox's comment above, it sounds like this is not intended behavior.  Then again, @sds's comment in #609 seems to go the other direction.

At this point, things are a little bit unclear to me... but imho, it is not very intuitive that you can get a `ColorVariable` warning when you're defining a color variable, simply because the color was specified with `rgba` instead of in hex.
